controller/volume_controller.go

```diff
@@ -1141,10 +1142,120 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
 		if oldState != v.Status.State {
 			vc.eventRecorder.Eventf(v, v1.EventTypeNormal, EventReasonAttached, "volume %v has been attached to %v", v.Name, v.Status.CurrentNodeID)
 		}
+
+		keepLocalReplicaPolicy, err := vc.ds.GetSettingAsBool(types.SettingNameKeepLocalReplicaOnSameNodeAsEngine)
+		if err != nil {
+			return err
+		}
+		if keepLocalReplicaPolicy {
+			if err := vc.handleKeepLocalReplicaOnSameNodeAsEngine(v, e, rs); err != nil {
+				return err
+			}
+		}
 	}
 	return nil
 }
 
+func (vc *VolumeController) handleKeepLocalReplicaOnSameNodeAsEngine(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica) error {
+	if e == nil {
+		return fmt.Errorf("BUG: handleKeepLocalReplicaWithEngine needs a valid engine")
+	}
+
+	if !vc.isSafeToModifyReplica(v, e, rs) {
+		return nil
+	}
+
+	// hasLocalReplica indicate whether there exist a replica on the same node as engine
+	hasLocalReplica := false
+	for _, r := range rs {
+		if r.Spec.NodeID == e.Spec.NodeID {
+			hasLocalReplica = true
+		}
+	}
+
+	if !hasLocalReplica {
+		if _, err := vc.addLocalReplicaOnSameNodeAsEngine(v, e, rs); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// check if there are have extra healthy replicas
+	healthyCount := 0
+	for rName, mode := range e.Status.ReplicaModeMap {
+		if _, exist := rs[rName]; exist && mode == types.ReplicaModeRW {
+			healthyCount++
+		}
+	}
+
+	if healthyCount <= v.Spec.NumberOfReplicas {
+		return nil
+	}
+
+	// Randomly delete extra non-local healthy replicas
+	// Sometime handleKeepLocalReplicaOnSameNodeAsEngine() is called more than once with the same input (v,e,rs).
+	// To make the deleting operation idempotent and prevent deleting more replica than needed,
+	// we always delete the replica with the smallest name. This doesn't destroy the randomness nature of the
+	// deleting operation because replica's name contains a random hashcode.
+	rNames := make([]string, 0, len(rs))
+	for rName := range rs {
+		rNames = append(rNames, rName)
+	}
+	sort.Strings(rNames)
+
+	for _, rName := range rNames {
+		r := rs[rName]
+		if r.Spec.NodeID != e.Spec.NodeID {
+			return vc.deleteReplica(r, rs)
+		}
+	}
+
+	return nil
+}
+
+// return true if successfully added, false otherwise
+func (vc *VolumeController) addLocalReplicaOnSameNodeAsEngine(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica) (bool, error) {
+	// Create a new replica struct. We don't create a replica in DS using createReplica() directly
+	// because we may need to delete the new replica if it fails to ScheduleReplicaToNode.
+	// This prevent UI from repeatedly show creating/deleting the new replica
```

Comment:
Do you mean a scheduling failed replica will be created each time when the volume is attached?

Reply:
Yeah. It is kind of. 

Assume that the local node is not available for scheduling replica. When the users turn on `KeepLocalReplicaOnSameNodeAsEngine`, Longhorn will repeatedly try to migrate a local replica.  

If we create a replica in DS using createReplica() directly, the following are repeatedly run:
1. create a new replica
2. schedule the new replica to the local node
3. if fail to schedule (in this case, it will fail because we assume local node is not available for scheduling), we have to delete the newly created replica because if we don't `ReconcileVolumeState` will schedule it to some other node latter on.

Step 1 and 3 cause a problem on Longhorn UI. The UI will show new replica immediately after step 1. Then UI deletes it immediately after step 3. The whole process is repeated again and again and users will see flashing on UI.

We can avoid this UI problem by not creating a replica in DS using createReplica() directly in step 1. Create an in-memory replica struct instead. As the result we don't have to do step 3 when fail to schedule.

Reply:
Got it.